tests/analyses/decompiler/test_decompiler.py

```diff
@@ -2898,6 +2898,131 @@ def test_multistatementexpression_od_read_char(self, decompiler_options=None):
+        # XXX: this a hack that should be fixed in some other place
+        text = d.codegen.text.replace("4294967295", "-1")
+
+        # first if-stmt should be a single scope with a return.
+        good_if_return = re.search("if \\(.*?\\)\n {8}return -1;", text)
+        assert good_if_return is not None
+
+        first_if_location = text.find("if")
+        assert first_if_location != -1
+        assert first_if_location == good_if_return.start()
+
+    @structuring_algo("phoenix")
+    def test_ifelseflatten_clientloop(self, decompiler_options=None):
+        bin_path = os.path.join(test_location, "x86_64", "decompiler", "clientloop.o")
+        proj = angr.Project(bin_path, auto_load_libs=False)
+        cfg = proj.analyses.CFGFast(normalize=True, data_references=True)
+
+        f = proj.kb.functions["client_request_tun_fwd"]
+        proj.analyses.CompleteCallingConventions(cfg=cfg, recover_variables=True, analyze_callsites=True)
+        d = proj.analyses[Decompiler](f, cfg=cfg.model, options=decompiler_options)
+
+        self._print_decompilation_result(d)
+        text = d.codegen.text
+
+        # find all ifs
+        all_if_stmts = list(re.finditer("if \\(.*?\\)", text))
+        assert all_if_stmts is not None
+        assert len(all_if_stmts) >= 2
+
+        # first if-stmt should be a single scope with a return.
+        first_good_if = re.search("if \\(.*?\\)\n {8}return 0;", text)
+        assert first_good_if is not None
+        assert first_good_if.start() == all_if_stmts[0].start()
+
+        # the if-stmt immediately after the first one should be a true check on -1
+        second_good_if = re.search("if \\(.*? == -1\\)", text)
+        assert second_good_if is not None
+        assert second_good_if.start() == all_if_stmts[1].start()
+
+    @structuring_algo("phoenix")
+    def test_ifelseflatten_certtool_common(self, decompiler_options=None):
+        bin_path = os.path.join(test_location, "x86_64", "decompiler", "certtool-common.o")
+        proj = angr.Project(bin_path, auto_load_libs=False)
+        cfg = proj.analyses.CFGFast(normalize=True, data_references=True)
+
+        f = proj.kb.functions["cipher_to_flags"]
+        proj.analyses.CompleteCallingConventions(cfg=cfg, recover_variables=True, analyze_callsites=True)
+        d = proj.analyses[Decompiler](f, cfg=cfg.model, options=decompiler_options)
+
+        self._print_decompilation_result(d)
+        text = d.codegen.text
+
+        # If any incorrect if-else flipping occurs, then there will be an if-stmt inside an if-stmt.
+        # In the correct output, there should only ever be 2 scopes (the function, and a single if-scope) of
+        # deepness in the full function. To verify this, we check that no scope of 3 deepness exists.
+
+        scope_prefix = "    "
+        bad_scope_prefix = scope_prefix * 3
+
+        assert scope_prefix in text
+        assert bad_scope_prefix not in text
+
+        # TODO: fix me, this is a real bug
```

Comment:
This TODO is real and is an if-stmt flipping bug. See the last if-stmt in the if-else chain.